### DISCUSSION TITLE
Use Mailpiler in combination with Dockermailserver

### DISCUSSION THREAD

**Discussion by DerGary** [docker-mailserver/docker-mailserver#3960] [Q&A]
My question is wether it is possible to configure Dockermailserver so that I can use mailpiler to archive everything incoming and outgoing email from one or multiple mail accounts from my dockermailserver instance. I already setup mailpiler from these resources: https://bitbucket.org/jsuto/piler/src/master/docker/ and dockermailserver. But I struggle on how to configure dockermailserver so that it sends every mail for a specific inbox to the mailpiler smtp server.

**Comment by glime-ai** [comment]
**Please note** that this response was created using generative AI, leveraging previously available information from the repository. This information may be helpful. However, it is our strong recommendation that you carefully consider the information before blindly applying it.

---

To configure Docker Mailserver (DMS) to send every email for a specific inbox to the Mailpiler SMTP server, you would typically use a feature of the mail transfer agent (MTA) used by DMS, which is Postfix. Postfix allows for the configuration of "always_bcc" settings, which can be used to send a copy of every email to a specified address. However, based on the provided context from the Docker Mailserver documentation and related resources, there isn't a direct mention of integrating with Mailpiler or setting up an "always_bcc" configuration specifically for Docker Mailserver.

Given this, the general approach to achieve your goal involves configuring Postfix within Docker Mailserver to add an "always_bcc" setting pointing to your Mailpiler SMTP server. This would involve customizing the Postfix configuration files used by Docker Mailserver. Here's a high-level overview of steps you might take, based on the general capabilities of Docker Mailserver for customization:

1. **Identify the Postfix Configuration File**: Docker Mailserver allows for advanced customization through providing "monkey-patching" configuration files (as mentioned in the Docker Mailserver introduction). For Postfix, this could involve the `postfix-main.cf` file.

2. **Add the `always_bcc` Setting**: You would add a line to the `postfix-main.cf` or a custom configuration file for Postfix, specifying the `always_bcc` option. For example:
   ```
   always_bcc = archive@your-mailpiler-address.com
   ```
   This setting tells Postfix to send a blind carbon copy (BCC) of all emails to the specified address, which would be your Mailpiler SMTP server's address.

3. **Apply the Configuration**: Depending on how Docker Mailserver is set up to handle custom configurations, you might need to mount your custom configuration file into the Docker container or use environment variables if supported for this specific setting. The documentation suggests that simple customizations can be supported through Docker Compose configuration and the `env-mailserver` configuration file, but more advanced customizations like this might require directly modifying or extending the configuration files.

4. **Restart Docker Mailserver**: After applying your configuration changes, restart the Docker Mailserver container to ensure the new settings take effect.

Please note, this approach assumes familiarity with Docker, Docker Compose, and the ability to customize Docker container deployments. Since the direct integration or specific configuration for Mailpiler is not detailed in the provided documentation, this solution is based on general Postfix functionality and Docker Mailserver's support for custom configurations.

For exact steps, including how to properly mount or modify configuration files within Docker Mailserver, refer to the Docker Mailserver documentation on customization and Postfix's documentation on the `always_bcc` setting.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=0d5b86fb7eea4a54acac405212c29aba&utm_medium=github).

**Comment by polarathene** [comment]
> But I struggle on how to configure dockermailserver so that it sends every mail for a specific inbox to the mailpiler smtp server.

Perhaps a bit of an example of talking through how you'd want the two to work?

If you want all mail for `@example.com` to be sent to this other mail server, then you could use an wildcard / catchall alias for that domain which points to a specific address that you want the mail sent to. I think that can be an IP + port if configured directly in `postfix-virtual.cf` (_I've not tried it, so it may not be compatible with the DMS config support, if not it can still be done with extra Postfix config and our `user-patches.sh` feature_).

**Reply by DerGary** [reply]
okay here is some more explanation: my setup is that i run a server with docker which contains both the DMS and mailpiler. Every email send to my server should be processed by DMS. If the mail that comes in is to be delivered to the mail account example@example.com then it should also be delivered/copied to the smtp server of mailpiler which is available at localhost with another port (as the default port is already allocated for DMS). Mailpiler then archives all the mails send to it. Additionally every email send from the account example@example.com must also additionally be send to the mailpiler smtp server for archiving.

**Reply by polarathene** [reply]
Awesome thanks!

If it's for a specific account you should be ok to use an alias like I mentioned previously.

So `user@example.com` can alias to itself (deliver to DMS mail account) and to your mailpiler container equivalent address. Presently DMS doesn't support  via `setup alias add` an alias that is an already existing mail account, but a workaround is to edit `postfix-virtual.cf` manually. You can find an existing open issue at our github repo to subscribe to and find more details on that workaround (I'm on phone atm so linking is a bit awkward).

In another discussion topic there was a user with a similar need to yours, sending the mail to the target destination but also sending a copy for archiving. We identified another alternative that configured an automatic bcc iirc, but I think there was some caveats to that.

If you can identify a good way to configure this with postfix we can try to support that properly in DMS. Aliases seem to be a good approach.

I think you don't have to worry about the port 25 concern, if you ensure both containers are on the same docker network, you can assign a hostname to the mailpiler container as well, it doesn't need to be a FQDN, so you could use `--hostname mail-archiver` and no port needs to be published. Then the alias will be like `user@example.com 
  user@example.com, user@mail-archiver` I think, which should work.

**Comment by DerGary** [comment]
I tried the suggestions from @polarathene It seemed to work but mailpiler does strange things and is badly documented so I will not spend more time on it.
